商品購入機能

thr3eisl4nd opened the pull request:
# WHAT
商品購入機能の実装

# WHY
出品されている商品を購入できるようにするため
出品した商品を販売できるようにするため

#### GyazoURL
- 必要な情報を適切に入力して「購入」ボタンを押すと、商品の購入ができる動画
https://gyazo.com/5c6fdbec5b4157e207cdd8f61c7f3976
- 入力に問題がある状態で「購入」ボタンが押された場合、情報は受け入れられず、購入ページでエラーメッセージが表示される動画
https://gyazo.com/c000bf6d0b6452a32a3d449cfcb76afd
- ログイン状態の場合でも、URLを直接入力して自身が出品していない売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移する動画
https://gyazo.com/5efffc736bfd056ec4cff2496eb6719e
- ログイン状態の場合でも、URLを直接入力して自身が出品した商品の商品購入ページに遷移しようとすると、商品の販売状況に関わらずトップページに遷移する動画
https://gyazo.com/d746c292197d9d82eaafffa49824121a
- ログアウト状態の場合は、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移する動画
https://gyazo.com/83fb6c9e150814cb6ae1bc7864d29993
https://gyazo.com/442b025f682cdb980fc1565f02d6a954
- 売却済みの商品は、画像上に「sold out」の文字が表示される動画（商品一覧機能実装時に未実装であった場合）
https://gyazo.com/a113fd4d951e62222ae2d2fc91c41f52
- 売却済みの商品は、画像上に「sold out」の文字が表示される動画（商品詳細機能実装時に未実装であった場合）
https://gyazo.com/359f2f9c7703bd45376eeb467b523e6f
- ログイン状態の場合でも、売却済みの商品には、「商品の編集」「削除」「購入画面に進む」ボタンが表示されない動画（商品詳細機能実装時に未実装であった場合）
https://gyazo.com/f6171302380baa3d356938a31f30ffc0
https://gyazo.com/5e39e4a206558cf8422f48197a27426a
- ログイン状態の場合でも、URLを直接入力して自身が出品した売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画（商品情報編集機能実装時に未実装であった場合）
https://gyazo.com/72ba0269072cd73077307ea7ccad6e45
- テスト結果の画像(最新のもの)
https://gyazo.com/a21a0a58c254a1e5d0ffaf2eade13af4

### 1/14(金) 12:34 修正しました。ご確認よろしくお願いいたします。

thr3eisl4nd committed:
Formオブジェクトの作成

order_addressファイルを作成、クラスを定義
Active::Modelを呼び出し、form_withメソッドに対応する機能とバリデーションを行う機能
attr_accessorを用いて、保存したい複数のテーブルのカラム名を全て扱えるようにする
バリデーションの処理を書く
データをテーブルに保存する処理を書く

thr3eisl4nd committed:
購入機能に必要なモデルやコントローラーを作成

購入機能に必要なモデルを用意(Order)
発送先情報に必要なモデルも作成(Address)、購入機能に必要なモデルとのアソシエーションを設定
購入機能に関するモデルと、商品情報に関するモデルとのアソシエーションを記述する
購入機能に必要なコントローラーを用意(orders)
購入機能に必要なルーティングを設定※ネストする
購入機能に必要なコントローラーへFormオブジェクトのインスタンスを生成する記述

thr3eisl4nd committed:
購入のためのフォームを実装

ordersのindexビューにurlオプションに指定のパスを記述
modelオプションにインスタンス変数を定義する

thr3eisl4nd committed:
フォームの情報を保存できるようにする/エラー文の表示

createアクションで処理が呼び出され、意図するパラメーターが送信されてるか確認
送られてきたパラメーターを保存する処理の記述

thr3eisl4nd committed:
テストコードを実装

thr3eisl4nd committed:
JavaScriptでtokenを生成してサーバーサイドへ送信する

①ブラウザのカード情報を取得
購入ボタンを押した際にイベントが発火
カード情報入力フォームのDOMがJS側で取得
取得したDOMからフォームに入力した値がJS側で取得出来ている

thr3eisl4nd committed:
公開鍵と共にカード情報をPAY.JP側へ送る

ブラウザのカード情報取得
公開鍵と共にカード情報をPAY.JP側へ送る
トークンが発行されたことを確認
トークンをサーバーサイドへ送る

thr3eisl4nd committed:
決済機能を実装

コントローラーへトークンが正常に送られていることを確認
コントローラーからPAY.JPへトークンを送信
決済が完了したことを確認

thr3eisl4nd committed:
鍵情報を環境変数として設定

thr3eisl4nd committed:
テストコードを編集する

thr3eisl4nd committed:
購入済み商品の条件分岐など

「ログイン状態の場合でも、自身が出品していない売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移すること」
「ログイン状態の場合でも、自身が出品した商品の商品購入ページに遷移しようとすると、商品の販売状況に関わらずトップページに遷移すること」
「ログアウト状態の場合は、商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移すること」

売却済みの商品は、画像上に「sold out」の文字が表示されるようになっていること（商品一覧機能）。
売却済みの商品は、画像上に「sold out」の文字が表示されるようになっていること（商品詳細機能）。
ログイン状態の場合でも、売却済みの商品には、「商品の編集」「削除」「購入画面に進む」ボタンが表示されないこと（商品詳細機能）。
ログイン状態の場合でも、自身が出品した売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移すること（商品情報編集機能）。

thr3eisl4nd committed:
rubocopでの確認/修正

zonodance commented on spec/factories/order_addresses.rb:
```diff
@@ -0,0 +1,13 @@
+    address2 { '東尾久' }
+    building_name { 'AK4801' }
+    phone_number { '05012345678' }
+    item_id { 1 }
+    user_id { 1 }
```
現在問題なく動いているかと思いますが、ユーザー_idと商品_idデータの作成方法が好ましい形でないため、修正をお願いします。

【理由】
外部キーに直接数字を入れるのは正しい情報が入っているかが分からないためです。
ユーザー_idと商品_idは外部キーであり、それぞれユーザーテーブルや商品テーブルで生成されたidが入ります。
user_id {2}と記載しても本当にそのレコードが存在するか分からないため、データの信頼性が低くなってしまう為です。

【改善案】
specファイルにて、値を生成する形にしましょう。
以下を参考に修正のチャレンジをお願いします。

フォームオブジェクト_spec.rbにユーザーと商品で作成したFactoryBotを使用してそれぞれのデータを生成しましょう。

```ruby
【例】
before do
①ユーザー = FactoryBot.create(:ユーザー)
②商品 = FactoryBot.create(:商品)
③@フォームオブジェクト = FactoryBot.build(:フォームオブジェクト, ユーザー_id: ユーザー.id , 商品_id: 商品.id)

④sleep(1)
今回beforeアクション内にデータが多い為、Mysqlエラーが出てしまう可能性があります。エラーが出た場合はsleepメソッドを使用してrubyをストップしている間に@order_addressが作れるようにしましょう。
(1)は1秒間という意味です。PCのスペックによっては足りない場合もございますので、2,3,と増やして試してみて下さい。
end
```

ご不明な点がありましたら、コネクトにてメンターにご相談ください！

zonodance commented on spec/models/order_address_spec.rb:
```diff
@@ -0,0 +1,62 @@
+        @order_address.valid?
+        expect(@order_address.errors.full_messages).to include('Zipcode Input correctly')
+      end
+      it '都道府県が必須であること' do
+        @order_address.shipment_source_id = ''
```
```ruby
@order_address.shipment_source_id = ""
```

こちら  @order_address.shipment_source_id = 「---が選択されているid」と記述いたしましょう！
prefecture_idカラムはinteger型で設計しているかと思われます！

zonodance commented on spec/models/order_address_spec.rb:
```diff
@@ -0,0 +1,62 @@
+        @order_address.token = nil
+        @order_address.valid?
+        expect(@order_address.errors.full_messages).to include("Token can't be blank")
+      end
```
以下のテストも追加しましょう！
order_address.rbにでuser_idとitem_idをpresence: trueに追加されている為、購入のテスト項目で以下のテストを行い確認してあげましょう！

```ruby
・user_id（購入者）が空だと購入できない
・item_id（購入商品）が空だと購入できない
```

zonodance commented on spec/models/order_address_spec.rb:
```diff
@@ -0,0 +1,62 @@
+        @order_address.valid?
+        expect(@order_address.errors.full_messages).to include("Phone number can't be blank")
+      end
+      it '電話番号は10桁以上11桁以内の半角数値のみ保存可能なこと' do
```
電話番号の異常系のテストコードは「9桁以下の場合」「12桁以上」と分けて記述しましょう！

zonodance reviewed (commented):
レビュー依頼ありがとうございます！
４点コメントさせていただきましたので、修正のほどよろしくお願いいたします。

thr3eisl4nd committed:
レビュー後の修正1.0

zonodance reviewed (commented):
レビュー依頼ありがとうございます！
実装もほぼ終了に近づいているかと存じております。

テストコードをご修正いただいたので、最新のテストコードの画像をこちらのプルリクエストに添付いただけますでしょうか。
正確なレビューのためによろしくお願いいたします。

thr3eisl4nd commented:
@zonodance 
ご返信ありがとうございます。
テストコードの結果の画像、最新のものに反映しております。
お手数ですが、もう一度ご確認いただいてもよろしいでしょうか。

sh10n7 commented on app/javascript/card.js:
```diff
@@ -0,0 +1,37 @@
+    Payjp.createToken(card, (status, response) => {
+      if (status == 200) {
+        const token = response.id;
+        console.log(token)
```
こちらのconsole.logは削除しましょう。

理由：
開発段階終了後にデバック用コードは明らかに不要であると判断できるためです。

sh10n7 commented on app/models/order_address.rb:
```diff
@@ -0,0 +1,18 @@
+class OrderAddress
+  include ActiveModel::Model
+  attr_accessor :zipcode, :shipment_source_id, :address1, :address2, :building_name, :phone_number, :order_id, :item_id,
```
```
:order_id
```
こちらは削除しましょう。

理由：
saveメソッド内でcreateするためです。

sh10n7 commented on spec/models/order_address_spec.rb:
```diff
@@ -0,0 +1,80 @@
+      it '電話番号は12桁以上の場合は登録出来ないこと' do
+        @order_address.phone_number = '080123456789'
+        @order_address.valid?
+        expect(@order_address.errors.full_messages).to include('Phone number Input only number')
```
異常系のテストの要件を満たしていることを確認するために、以下のテストも実施しましょう。

```
■電話番号 
・電話番号に半角数字以外が含まれている場合は購入できない
```

sh10n7 reviewed (commented):
レビュー依頼ありがとうございます。
修正点確認させていただきました。

追加で3点修正をお願いしております。ご確認いただけますと幸いです。
​
修正頂いた後、コミットプッシュを確認いただき、再度レビュー依頼のご提出をお願い致します！
その際、**修正後のテスト実行結果の画像**も添付いただきますようお願いいたします。

ご不明点や疑問点ございましたら、通話メンターにご相談ください！
どうぞ、よろしくお願いいたします。

thr3eisl4nd committed:
レビュー後の修正1.1

jsファイルのconsole.log削除
attr_accessor内の:order_id削除
テストにて「電話番号半角数値以外が含まれている場合は購入できない」を追加

chobi555 commented:
レビュー依頼ありがとうございます！
LGTMとさせていただきます、実装お疲れさまでした！

マージ後、ローカルのmasterブランチに忘れずにプルしましょう。
その後、新しいブランチを作成し次の実装に取り組んでいただけますと幸いです。
引き続き頑張っていきましょう！
